src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,196 @@
+        T oldItem = null;
+        if (foundedNode != null) {
+            oldItem = foundedNode.item;
+            foundedNode.item = value;
+        }
+        return oldItem;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexOfBoundsException(index);
+        Node<T> foundedNode = getNodeByIndex(index);
+        if (unlinkIfObjectHead(foundedNode.item)) {
+            return foundedNode.item;
+        }
+        unlink(foundedNode);
+        return foundedNode.item;
     }
 
     @Override
     public boolean remove(T object) {
+        if (size == 0) {
+            return false;
+        }
+        if (unlinkIfObjectHead(object)) {
+            return true;
+        }
+        Node<T> foundedNode = getNodeByValue(object);
+        if (foundedNode != null) {
+            unlink(foundedNode);
+            return true;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
+        return size == 0;
+    }
+
+    private void checkIndexOfBoundsException(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index " + index + " doesn't exist");
+        }
+    }
+
+    private void checkAddIndexOfBoundsException(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index " + index + " out of bound.");
+        }
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        if (index < size / 2 || index < 2) {
+            return searchFromHead(index);
+        } else {
+            return searchFromTail(index);
+        }
+    }
+
+    private Node<T> searchFromHead(int index) {
+        Node<T> currentNode = head;
+        for (int i = 0; i < index; i++) {
+            currentNode = currentNode.next;
+        }
+        return currentNode;
+    }
+
+    private Node<T> searchFromTail(int index) {
+        Node<T> currentNode = tail;
+        for (int i = size - 1; i > index; i--) {
+            currentNode = currentNode.prev;
+        }
+        return currentNode;
+    }
+
+    private Node<T> getNodeByValue(T value) {
+        Node<T> currentNode = head;
+        while (!ObjectService.equals(currentNode.item, value)) {
+            if (currentNode.next == null) {
+                return null;
+            }
+            currentNode = currentNode.next;
+        }
+        return currentNode;
+    }
+
+    private void unlink(Node<T> node) {
+        node.prev.next = node.next;
+        node.next.prev = node.prev;
+        size--;
+    }
+
+    private boolean unlinkIfObjectHead(T object) {
+        if (ObjectService.equals(head.item, object)) {
+            if (head.next != null) {
+                head.next.prev = head.prev;
+            }
+            head = head.next;
+            size--;
+            return true;
+        }
         return false;
     }
```

Comment:
In my opinion unlink should processed tail, head, mid cases in one method. Also what if u need unlink tail?